src/coreclr/jit/assertionprop.cpp

```diff
@@ -3551,6 +3551,86 @@ GenTree* Compiler::optAssertionProp_BlockStore(ASSERT_VALARG_TP assertions, GenT
     return nullptr;
 }
 
+//------------------------------------------------------------------------
+// optAssertionProp_ModDiv: Convert DIV/MOD to UDIV/UMOD if both operands
+//    can be proven to be non-negative, e.g.:
+//
+//    if (x > 0)         // creates "x > 0" assertion
+//        return x / 8;  // DIV can be converted to UDIV
+//
+// Arguments:
+//    assertions - set of live assertions
+//    tree       - the DIV/MOD node to optimize
+//    stmt       - statement containing DIV/MOD
+//
+// Returns:
+//    Updated UDIV/UMOD node, or "nullptr"
+//
+GenTree* Compiler::optAssertionProp_ModDiv(ASSERT_VALARG_TP assertions, GenTreeOp* tree, Statement* stmt)
+{
+    if (optLocalAssertionProp || !varTypeIsIntegral(tree) || BitVecOps::IsEmpty(apTraits, assertions))
+    {
+        return nullptr;
+    }
+
+    // For now, we're mainly interested in "X op CNS" pattern (where CNS > 0).
+    // Technically, we can check assertions for both operands, but it's not clear if it's worth it.
+    if (!tree->gtGetOp2()->IsNeverNegative(this))
+    {
+        return nullptr;
+    }
+
+    const ValueNum dividendVN = vnStore->VNConservativeNormalValue(tree->gtGetOp1()->gtVNPair);
+    for (AssertionIndex index = 1; index <= optAssertionCount; index++)
```

Comment:
For the most part we're probably better of iterating through the bits of `assertions` here, as hopefully that iteration is cheap (using HW accelerated `bsf` or similar) and that most bits in the vector are *not* set.

I am tempted to convert most of the search loops like this in assertion prop over to bitvector driven loops.

Reply:
Do we already have an example where we do that? I didn't bother because TP impact was zero 🙂 (because DIV/MOD aren't too common)

Reply:
Replaced with BitVecOps::Iter assuming that what you meant